src/sync/sync_session.cpp

```diff
@@ -413,10 +412,36 @@ std::string SyncSession::get_recovery_file_path()
                                           util::create_timestamped_template("recovered_realm"));
 }
 
+void SyncSession::update_error_and_mark_file_for_deletion(SyncError& error, ShouldBackup should_backup)
+{
+    // Add a SyncFileActionMetadata marking the Realm as needing to be deleted.
+    std::string recovery_path;
+    auto original_path = path();
+    error.user_info[SyncError::c_original_file_path_key] = original_path;
+    if (should_backup == ShouldBackup::yes) {
+        recovery_path = get_recovery_file_path();
+        error.user_info[SyncError::c_recovery_file_path_key] = recovery_path;
+    }
+    auto action = ((should_backup == ShouldBackup::yes)
+                   ? SyncFileActionMetadata::Action::BackUpThenDeleteRealm
+                   : SyncFileActionMetadata::Action::DeleteRealm);
+    SyncManager::shared().perform_metadata_update([this,
+                                                   action,
+                                                   original_path=std::move(original_path),
+                                                   recovery_path=std::move(recovery_path)](const auto& manager) {
+        SyncFileActionMetadata(manager,
```

Comment:
It's surprising to me that creating a C++ object like this modifies a Realm file on disk. It took me a bit of reading to figure out what was going on here.